### PR TITLE
[813] Admin can add contact types to an organization

### LINKED ISSUE #813: Admins can add contact_types and groups
Part of epic #4 // relates to issues #815, #816 

**What type of user is this for? [admin/supervisor/volunteer/all or All CASA Admin]**
admin

**Context**
`contact_types` associated with `case_contacts` vary from CASA Org to CASA Org. Giving admins the ability to add `contact_types` associated with their unique Org instance will increase accuracy in reporting.
Grouping `contact_types` will help paint a higher level picture of which `contact_types` are being contacted most and least.

**How to access on the frontend in QA:**
<a href="https://casa-qa.herokuapp.com/">Log in to QA</a> and use these credentials to access the admin dashboard:
username – casa_admin1@example.com
password – 123456

**Where does/should this occur?**
On the <a href="https://casa-qa.herokuapp.com/casa_orgs/2/edit">Edit CASA Org page</a> in the admin dashboard.

**Frontend**
Admins should be able to add `contact_types` and `contact_groups` associated with `case_contacts` to be used in their organization's instance and display on create and edit case contact forms. 

**Backend**
_Current structure:_
`case_contact: … contact_types(arr(str))`
_Future structure should look something like this:_ 
`case_contact: …
contact_type: casa_org_id, name, group
join table (one(case_contact)-to-many(contact_type)): case_contact_contact_type: case_contact_id, contact_type_id uniq(case_contact_id, contact_type_id)`

### PR DESCRIPTION
**Pull request by gmfvpereira**
### What github issue is this PR for, if any?
Resolves #813

### What changed, and why?
Contact types were being saved on `case_contacts.contact_types` column as a list of strings.
This PR creates 3 tables ContactTypes, ContactTypeGroups, and CaseContactContactType (joining table) to store the same data but with the ability to have a different list of contact types for each Casa Organization.

Other than creating the tables this PR also migrates existing data to this new structure and updates the reads and writes to keep it up to date.

### How will this affect user permissions?
It won't affect it.

### How is this tested? (please write tests!) 💖💪
We need to test:
- Creating/Editing a CaseContact are saving the contact types properly
- Reports are including correct types
- Adding Contact Types and Groups
### Screenshots please :)
Screen Shot 2020-09-24 at 1 14 12 PM
Screen Shot 2020-09-24 at 1 14 20 PM
Screen Shot 2020-09-24 at 1 14 26 PM
Screen Shot 2020-09-24 at 1 14 32 PM

### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
`alt text`

### CODE REVIEW

**Comment by gmfvpereira** [comment]
**PLEASE READ**
There's a problem with this PR. During deployment, we may miss some recently added data:

1 - Production deployment is started
2- Migrations start 
3- Migrations finish, but the deployment is not finished yet, so the code that update/read from the new tables is not running yet.
4- A user creates a `CaseContact` and the contact type gets saved on `case_contact.contact_type` column BUT NOT saved on the new tables.
5- Deployment finish.
6- If we check that CasaContact created on step 4, it won't show any ContactType.

Is this a problem @Zrrrpy ? 
If so, how do you usually handle this kind of situation @jmkoni ?

File: app/controllers/contact_type_groups_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  before_action :must_be_admin
+
+  def new
+    @contact_type_group = ContactTypeGroup.new
+  end
+
+  def create
+    @contact_type_group = ContactTypeGroup.new(contact_type_group_params.merge(casa_org: current_organization))
+
+    respond_to do |format|
```
Comment (drborges) [inline app/controllers/contact_type_groups_controller.rb:11]: Do we need the `respond_to` block even though we are dealing with `HTML` only?

File: app/models/case_contact.rb [hunk 2]
```diff
@@ -38,8 +42,10 @@ class CaseContact < ApplicationRecord
     where(want_driving_reimbursement: want_driving_reimbursement) if want_driving_reimbursement == true || want_driving_reimbursement == false
   }
   scope :contact_type, ->(contact_type = nil) {
-    where("? = ANY (contact_types)", contact_type) if contact_type.present?
+    joins(:db_contact_types)
+      .where("contact_types.name in (?)", contact_type) if contact_type.present?
```
Comment (drborges) [inline app/models/case_contact.rb:46]: Certainly not an issue, but for consistency with how this has been done in other places:

```suggestion
      .where(contact_types: { name: contact_type }) if contact_type.present?
```

File: app/views/case_contacts/_form.html.erb [hunk 3]
```diff
@@ -35,15 +35,15 @@
   <div id="contact-type-form" class="field contact-type form-group">
     <h2 id="contact-type-label"><%= form.label :contact_types %></h2>
 
-    <% contact_types = CaseContact::CONTACT_TYPES.each_with_index do |contact_type, index| %>
+    <% @contact_types.each_with_index do |contact_type, index| %>
       <div class="form-check">
-        <%= check_box_tag "case_contact[contact_types][]",
-          contact_type,
-          case_contact.contact_types&.include?(contact_type),
+        <%= check_box_tag "case_contact[case_contact_contact_type_attributes][][contact_type_id]",
```
Comment (drborges) [inline app/views/case_contacts/_form.html.erb:40]: Is there a form helper we could use to simplify this?

File: db/schema.rb [hunk 4]
```diff
@@ -86,6 +95,22 @@
...
+    t.bigint "casa_org_id", null: false
+    t.string "name", null: false
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["casa_org_id"], name: "index_contact_type_groups_on_casa_org_id"
+  end
+
+  create_table "contact_types", force: :cascade do |t|
+    t.bigint "contact_type_group_id", null: false
+    t.string "name"
```
Comment (drborges) [inline db/schema.rb:108]: Is `name` really nullable?

**Review by drborges — APPROVED** [review]
This looks great, I could not spot anything that would raise a red flag, I've left a few notes but really things that don't necessarily have to be addressed and could very well be addressed in upcoming PRs.

Very nice work 👏🏼

File: app/views/case_contacts/_form.html.erb (hunk 3 above)
Comment (gmfvpereira) [reply app/views/case_contacts/_form.html.erb:40]: I tried a few helpers but I couldn't make them work. :sad:

I'm working on another PR now that have a similar form input, I'll give it another chance and if it end up making it work I'll update here.

File: app/controllers/contact_type_groups_controller.rb (hunk 1 above)
Comment (gmfvpereira) [reply app/controllers/contact_type_groups_controller.rb:11]: Whoops, left over.

File: db/schema.rb (hunk 4 above)
Comment (gmfvpereira) [reply db/schema.rb:108]: Not really, I'll fix it.

**Review by jmkoni — APPROVED** [review]

**Comment by gmfvpereira** [comment]
Thanks for the review folks!

While working on another story similar to this one I found an issue with the way I used `accept_nested_attributes_for` that is not allowing to remove a relationship between `CasaContact` and `ContactTypes`. The way it is it only allow to add contact types. I'm working on a fix.

File: db/migrate/20200922170308_link_case_contacts_to_contact_types.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  def change
+    create_table :case_contact_contact_types do |t|
+      t.references :case_contact, null: false
+      t.references :contact_type, null: false
+
+      t.timestamps
+    end
+
+    code_to_name_mapping = {
+      attorney: ContactType.where(name: "Attorney").first,
```
Comment (compwron) [inline db/migrate/20200922170308_link_case_contacts_to_contact_types.rb:11]: This caused problems - will be fixed in https://github.com/rubyforgood/casa/issues/1056

**Pull request merged**